File: Selection/C++/Selection.cpp
//
//  Selection.cpp
//  
//
//  Owen Gallagher & Rebecca Drucker
//  2/14/18
//
//

#include <iostream>
#include <math.h>

using namespace std;

int n = 10;
int* A;

int minIndex = 0;   //next smallest
int temp = 0;       //for swaps

void describe() {
    cout << "A: ";
    
    for (int i=0; i<n; i++) {
        cout << A[i] << " ";
    }
    
    cout << endl;
}

void selectionSort() {
    for (int i=0; i<n-1; i++) {
        //find next lowest
        int minIndex = i;
        for (int j=i+1; j<n; j++) {
            if (A[j] < A[minIndex]) {
                minIndex = j;
            }
        }
        
        //swap ith lowest with ith element
        temp = A[i];
        A[i] = A[minIndex];
        A[minIndex] = temp;
    }
}

int main() {
    A = new int[n];
    
    cout << "C++ selection sort of n=" << n << " elements." << endl;
    
    //fill A
    srand(time(nullptr));
    
    for (int i=0; i<n; i++) {
        A[i] = round(rand() % 100);
    }
    
    //unsorted
    describe();
    
    selectionSort();
    
    //sorted
    describe();
    
    return 0;
}

File: Selection/Java/Selection.java
/* 
    Rebecca Drucker & Owen Gallagher
    SelectionSort.java
 */

public class Selection {
    static int N = 10;
    static int A[] = new int[N];
    
    static int minIndex = 0;   //next smallest
    static int temp = 0;       //for swaps
    
    public static void describe() {
        System.out.print("A: ");
        for (int i = 0; i < N; i++){
            System.out.print(A[i] + " ");
        }
        System.out.println();
    }

    public static void selectionSort(){
        for (int i = 0; i < N - 1; i++){
            minIndex = i;
            for (int j = i + 1; j < N; j++) {
                if (A[j] < A[minIndex]){
                    minIndex = j;
                }
            }

            temp = A[i];
            A[i] = A[minIndex];
            A[minIndex] = temp;
        }
    }

    public static void main(String[] args){
        System.out.println("JAVA selection sort of n=" + N + " elements.");
        
        // set up A
        for (int i = 0; i < N; i++){
            A[i] = (int)(Math.random() * 100);
        }

        // print unsorted A
        describe();

        selectionSort();

        // print sorted A
        describe();
    }
}
